### ISSUE TITLE
WS2801 5050 LEDs flashing rapidly

### ISSUE THREAD

**Issue by jimmyeao** [waldobronchart/AmbientLightServer#4]
Hi Waldo,
I have replaced my original LEDs with 2 x 48 WS2801 5050 leds, as these are flatter and easier to mount
I have what appears to be a timing issue (?) it doesn't seem to matter how many of the LEDs I set it to use in HardwareConfig.h they just flash wildly, and the last 9 or so LEDs in the strip dont light up at all.
From reading around, Im guessing the signal is deteriorating, or these particular LEDs aren't compatible with whichever library you are using to drive the LEDs (Although I had a quick look at your code, seems you are going direct rather than using something like the Adafruit ws2801 library?)
Have tested the LEDs with Python, and this works fine - I'm stumped!
these are the LEDs I used if it helps:
https://www.amazon.co.uk/gp/product/B071122MMD/ref=oh_aui_detailpage_o02_s00?ie=UTF8&psc=1
 I should say this happens if I try and use a static color also.

Many thanks

**Comment by waldobronchart** [comment]
Any progress on this?

Some thoughts:
- The pins are not assigned properly; make sure to use the wiringPi GPIO layout for numbering, not the pi layout
- The LEDs at the end of the strand not lighting up may be exactly the same issue that I had, which is why I opted to update each strand separately. The signal seemed to deteriorate towards the end of the strand.
- Look up the documentation on the WS2801 and check that the code in LEDController::ShiftOut8Bits is correct (https://github.com/waldobronchart/AmbientLightServer/blob/master/LEDController.cpp#L118)

**Comment by jimmyeao** [comment]
Hi, thanks for the pointers. Happens from the 2nd led onwards, checked wiring (works with another set of leds, not surface mounted though)
I’ll play with the led controller as suggested and see if that helps. I do get drop off towards the end of the strips, I wonder how Hyperion gets around this issue, I see people use single strips of 96 on there?

**Comment by jimmyeao** [comment]
Had a look, cant see anything obvious. How would I change the timing to say, drop it down to 2 mhz?

**Comment by waldobronchart** [comment]
Maybe add some delay between clock 0 and 1?
```
digitalWrite(clockPin, 0);
digitalWrite(dataPin, val);
delay(1);
digitalWrite(clockPin, 1);
```

I'm just guessing here

**Comment by jimmyeao** [comment]
OK, some good news - delay didn't work, but `usleeep(10);` did. It slowed the FPS down to 6, but no noise at the end of the strips. WIll now play around and see how much of a delay I need before it becomes unstable again.
Thanks for the pointer :)

**Comment by jimmyeao** [comment]
Nailed it. 30FPS and no flicker, using 48 LEDS per strip.
usleep wasnt fine enough, so I used 
`nanosleep(100, NULL);`
and it appears to be working perfectly.

**Comment by waldobronchart** [comment]
Awesome, good to hear! Can you share the full version of your `LEDController::ShiftOut8Bits` function?

**Comment by jimmyeao** [comment]
```
void LEDController::ShiftOut8Bits(int clockPin, int dataPin, uint8_t c)
{
	#ifdef RASPBERRY_PI
	for (int bit = 0; bit < 8; bit++)
	{
		bool val = (c >> (7 - bit)) & 1;

		digitalWrite(clockPin, 0);
		digitalWrite(dataPin, val);
		nanosleep(100, NULL);
		digitalWrite(clockPin, 1);
	}
	#endif
```

**Comment by waldobronchart** [comment]
Great, changed my code too. Thanks for the investigation. I think I'll no longer need to update each strand separately.

**Comment by jimmyeao** [comment]
Now all I need to do is figure out how to exit gracefully from a key press - I’ve noticed if I just ctrl c out it takes a while before it releases the camera properly - if I run after quitting it doesn’t pick up the camera and repost 120+ fps

**Comment by Vovchick2004** [comment]
Gentlemen,
I have an issue with my LED strip where it works fine most of the time, but then it will go into a wild cycle of just flashing randomly. Sometimes for 0.5 sec to a few seconds of just flashing randomly.
I've played with different delays but it seems to still do the same.
I've checked the RGB output and that is not jumping, so I know it's the actual strip control.
I also get the same behavior with different lengths: started with 5M (160 LEDs) and now I'm down to 30 LEDs (1 side length) and it is still doing that. 
Any help will be appreciated.
